### ISSUE TITLE
implement pytorch's compliance.kaldi.fbank() in librosa?

### ISSUE THREAD

**Issue by AlexJian1086** [librosa/librosa#1379]
Hi everyone,
I would really appreciate if someone could let me know how to replicate pytorch's **compliance.kaldi.fbank()** function in **librosa?** I've gone through alot of literature and forums but haven't really found a way to replicate their parameters. Any help would be appreciated.

**Comment by bmcfee** [comment]
Short answer is this *might* be possible, but probably not for every combination of parameters.

I took a look at both the original Kaldi docs and the pytorch compatibility layer, and ... yikes, that's a lot of parameters!  The docs aren't entirely forthcoming about what they all mean and do, so that doesn't help.  From what I can tell, "fbank" here actually means a mel spectrogram with a bunch of pre-processing.

https://pytorch.org/audio/stable/compliance.kaldi.html#fbank lists the parameters. Let's take these in order:

- waveform: torch.Tensor,   - the input signal, nothing to do here.
-   blackman_coeff: float = 0.42; if you're not changing this, you can use `window='blackman'` and it will work.  (Scipy's blackman window uses 0.42, but does not expose a parameter to change it.) Otherwise, you'll have to use a custom window function, but that's easily supported by librosa's stft.
-   channel: int = -1,  - which axis to operate on, nothing to do here.
-   dither: float = 0.0, - if this stays at 0, you're okay.  We don't implement dithering though, so you won't be able to change it.
-   energy_floor: float = 1.0, - 
-   frame_length: float = 25.0, - this is in milliseconds, so you'll have to convert to samples manually to get `n_fft`
-   frame_shift: float = 10.0, - likewise but for `hop_length`
-   high_freq: float = 0.0, - this is supported by our mel filterbank (`fmax`, should be `sr/2` to match default behavior)
-   htk_compat: bool = False - they're using this just to determine where in the output to place the energy channel, and does not affect the same functionality that our `htk=` parameter does.
-   low_freq: float = 20.0, - also supported by our filterbank (`fmin`)
-   min_duration: float = 0.0, - unsupported / you'll have to truncate manually
-   num_mel_bins: int = 23, - translates to `n_mels`
-   preemphasis_coefficient: float = 0.97, - if you want this, you'll need to apply preemphasis before calling melspectrogram
-   raw_energy: bool = True, - not supported, I don't see an easy way to match this behavior with our implementation
-   remove_dc_offset: bool = True, - also not supported
-   round_to_power_of_two: bool = True, - I think this will be done implicitly since we decouple frame and window size, but haven't checked
-   sample_frequency: float = 16000.0, - `sr`
-   snip_edges: bool = True, - setting `center=True` ought to do this, or something close to it.
-   subtract_mean: bool = False, - this is just post-processing, so you could mean-subtract afterward
-   use_energy: bool = False, - likewise, you could tack on RMS as an extra dimension if you want, though you'll have to be careful to match parameters there.
-   use_log_fbank: bool = True, - This is confusing in the documentation, but what it actually means is log-scaling the filter response (features), not the filterbank itself.  Pytorch does `log(maximum(x, epsilon)`, we do something a bit fancier in `amplitude_to_db` but it should be easy enough to reimplement.
-   use_power: bool = True, - Should be equivalent to setting `power=2` (True) or `power=1` (False) in call to melspectrogram
-   vtln_high: float = -500.0, - These three are going to be tricky, as we don't implement any vocal tract length normalization.  I don't think it matters if `vtln_warp = 1.0`, but I haven't looked carefully.
-   vtln_low: float = 100.0,
-   vtln_warp: float = 1.0,
-   window_type: str = 'povey' - scipy doesn't implement povey, but it's easy enough to make a custom window function for this.  Pytorch just does `hann(...)**0.85`.


So aside from all that, the one other thing to be careful of is using `htk=True` for building mel filterbanks.  Our default (False) uses a different mapping of mels to frequencies, but setting htk=True will use the same frequency mapping as htk and pytorch/kaldi.

As I said above, you might be able to squeeze out some compatible features here, but it's going to be rather delicate and only work for some parameter settings.

**Comment by DEDSEC-Roger** [comment]
I think when `snip_edges = True`, we should set `center = False` to compatible with it. Following code is from `torchaudio.compliance.kaldi.fbank`, when `center = False`, n_frms = int((n_samples - win_length) / hop_length) + 1 in librosa.
```
def _get_strided(waveform: Tensor, window_size: int, window_shift: int, snip_edges: bool) -> Tensor:
    r"""Given a waveform (1D tensor of size ``num_samples``), it returns a 2D tensor (m, ``window_size``)
    representing how the window is shifted along the waveform. Each row is a frame.

    Args:
        waveform (Tensor): Tensor of size ``num_samples``
        window_size (int): Frame length
        window_shift (int): Frame shift
        snip_edges (bool): If True, end effects will be handled by outputting only frames that completely fit
            in the file, and the number of frames depends on the frame_length.  If False, the number of frames
            depends only on the frame_shift, and we reflect the data at the ends.

    Returns:
        Tensor: 2D tensor of size (m, ``window_size``) where each row is a frame
    """
    assert waveform.dim() == 1
    num_samples = waveform.size(0)
    strides = (window_shift * waveform.stride(0), waveform.stride(0))

    if snip_edges:
        if num_samples < window_size:
            return torch.empty((0, 0), dtype=waveform.dtype, device=waveform.device)
        else:
            m = 1 + (num_samples - window_size) // window_shift
    else:
        reversed_waveform = torch.flip(waveform, [0])
        m = (num_samples + (window_shift // 2)) // window_shift
        pad = window_size // 2 - window_shift // 2
        pad_right = reversed_waveform
        if pad > 0:
            # torch.nn.functional.pad returns [2,1,0,1,2] for 'reflect'
            # but we want [2, 1, 0, 0, 1, 2]
            pad_left = reversed_waveform[-pad:]
            waveform = torch.cat((pad_left, waveform, pad_right), dim=0)
        else:
            # pad is negative so we want to trim the waveform at the front
            waveform = torch.cat((waveform[-pad:], pad_right), dim=0)

    sizes = (m, window_size)
    return waveform.as_strided(sizes, strides)
```
